### ISSUE TITLE
perf: Newton Hessian is O(n²) autograd calls — impractical beyond ~100 parameters

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#68]
## Problem

Newton computes the Hessian by calling `grad()` once per row:

```python
for idx in range(g.shape[0]):
    H[idx] = torch.hstack([
        d.view(1, -1) if d is not None else ...
        for param, d in zip(params, grad(g[idx], params, ...))
    ])
```

For `n` parameters, this makes `n` separate autograd passes, each computing `n` partial derivatives — an O(n²) number of autograd calls. In practice, Newton is unusably slow beyond ~100 parameters.

For a numerical mathematician who might want to solve systems with thousands of parameters, this is a hard blocker.

## Proposal

Replace the row-by-row approach with one of:

1. **`torch.autograd.functional.hessian()`** — built-in, handles all edge cases, but returns a nested tuple structure that needs flattening.

2. **Vector-Jacobian product (VJP) approach** — compute the Hessian-vector products in a single pass using `torch.autograd.grad(g, params, grad_outputs=...)`, which is the standard efficient approach in the optimization literature.

3. **At minimum, document the scaling limitation** in the class docstring so users know what to expect.

Option 2 is the most "numerical mathematician-friendly" since it also enables Hessian-free Newton (Krylov methods) where you never form the full Hessian.

**Comment by kovacoj** [comment]
## Proposed fix

Replace the row-by-row Hessian construction with `torch.autograd.functional.hessian`, which computes the full Hessian in a single call:

```python
# Before (O(n) autograd calls)
H = prototype.new_empty(self.numel, self.numel)
for idx in range(g.shape[0]):
    if not g[idx].requires_grad:
        H[idx] = prototype.new_zeros(self.numel)
        continue
    H[idx] = torch.hstack([
        d.view(1, -1) if d is not None else param.new_zeros(1, param.numel())
        for param, d in zip(
            params,
            grad(g[idx], params, create_graph=True, retain_graph=True, allow_unused=True)
        )
    ])

# After (single autograd call)
from torch.autograd.functional import hessian as torch_hessian

def _flat_closure(flat_params):
    with torch.no_grad():
        load_flat_params_(params, flat_params)
    return closure()

H = torch_hessian(_flat_closure, flat_params(params))
```

**Pros:**
- Single autograd call instead of `n` — significantly faster for moderate `n`.
- Handles edge cases (unused parameters, zero gradients) correctly via PyTorch's own implementation.
- Returns a proper `(n, n)` tensor directly.

**Cons:**
- `torch.autograd.functional.hessian` requires a function of a flat tensor, so we need to wrap the closure.
- For very large `n` (>1000), even the single-call Hessian is too expensive (O(n²) memory). A Hessian-free Newton-Krylov approach would be needed, but that's a larger change.

**Migration path:**
1. **Short term:** Switch to `torch.autograd.functional.hessian` — immediate speedup, same correctness, no API change.
2. **Long term (optional):** Add a `hessian_free=True` option that uses Hessian-vector products (`grad(g, params, grad_outputs=v)`) with conjugate gradient, enabling Newton for large-scale problems without ever forming the full Hessian.

**Note on unused-parameter handling:** The current row-by-row code has special handling for `g[idx]` that doesn't require grad. `torch.autograd.functional.hessian` handles this naturally — unconnected entries are zero. We should verify this matches the current behavior in a test.
